rust/src/delta.rs

```diff
@@ -613,6 +662,175 @@ impl std::fmt::Debug for DeltaTable {
+    },
+
+    #[error("RecordBatch is missing partition column in Delta schema.")]
+    MissingPartitionColumn,
+
+    #[error("Storage interaction failed: {source}")]
+    Storage { source: StorageError, },
+
+    #[error("DeltaTable interaction failed: {source}")]
+    DeltaTable { 
+        #[from]
+        source: DeltaTableError,
+    },
+
+    #[error("Action serialization failed: {source}")]
+    ActionSerializationFailed {
+        #[from]
+        source: serde_json::Error, 
+    }
+}
+
+/// Error that occurs when a single transaction commit attempt fails
+#[derive(thiserror::Error, Debug)]
+pub enum TransactionCommitAttemptError {
+    #[error("Version already exists: {source}")]
+    VersionExists { source: StorageError },
+
+    #[error("Commit Failed due to DeltaTable error: {source}")]
+    DeltaTable {
+        #[from]
+        source: DeltaTableError,
+    },
+
+    #[error("Commit Failed due to StorageError: {source}")]
+    Storage { source: StorageError, }
+}
+
+impl From<StorageError> for TransactionCommitAttemptError {
+    fn from(error: StorageError) -> Self {
+        match error {
+            StorageError::AlreadyExists(_) => TransactionCommitAttemptError::VersionExists { source: error },
+            _ => TransactionCommitAttemptError::Storage { source: error },
+        }
+    }
+}
+
+/// Options for customizing behavior of a `DeltaTransaction`
+pub struct DeltaTransactionOptions {
+    /// number of retry attempts allowed when committing a transaction
+    max_retry_commit_attempts: u32,
+}
+
+impl DeltaTransactionOptions {
+    /// Creates a new `DeltaTransactionOptions`
+    pub fn new(max_retry_commit_attempts: u32) -> Self {
+        Self { max_retry_commit_attempts, }
+    }
+}
+
+const DEFAULT_DELTA_MAX_RETRY_COMMIT_ATTEMPTS: u32 = 10_000_000;
+
+/// Object representing a delta transaction
+pub struct DeltaTransaction<'a> {
+    delta_table: &'a mut DeltaTable,
+    options: DeltaTransactionOptions,
+}
+
+impl<'a> DeltaTransaction<'a> {
+    /// Creates a new delta transaction.
+    /// Holds a mutable reference to the delta table to prevent outside mutation while a transaction commit is in progress.
+    /// Transaction behavior may be customized by passing an instance of `DeltaTransactionOptions`.
+    pub fn new(delta_table: &'a mut DeltaTable, options: Option<DeltaTransactionOptions>) -> Self {
+        DeltaTransaction { 
+            delta_table,
+            options: options.map_or_else(|| DeltaTransactionOptions { 
+                max_retry_commit_attempts: DEFAULT_DELTA_MAX_RETRY_COMMIT_ATTEMPTS 
+            }, |o| o),
+        }
+    }
+
+    /// Commits the given actions to the delta log.
+    /// This method will retry the transaction commit based on the value of `max_retry_commit_attempts` set in `DeltaTransactionOptions`.
+    pub async fn commit_all(&mut self, actions: &[Action], _operation: Option<DeltaOperation>) -> Result<DeltaDataTypeVersion, DeltaTransactionError> {
```

Comment:
After thinking more this, I agree with @nevi-me that it would be handy to maintain the list of staged actions within the transaction struct itself. I can see for more complex transactions, users would do some non-trivia computation, produce some actions, then do so other non-trivia computations and produce other actions. It would be a mental burden for them to having to keep track of all the actions needed to commit at the very end v.s. being able to "queue" those actions as the flow of application logic goes on.

If we remain all the staged actions in an internal vector and add a `commit` method to transaction struct, then `commit_all` can be simplified to a helper method that adds more action to the vector and calls `commit`. So users can still use this interface for simple transactions.

I can see why you didn't want to maintain an internal vector though, it's extra work to deal with action struct clone or reference passing :P But it looks like it should be pretty straight forward to derive `Clone` for all the action types, then we don't need to deal with lifetime and references. Given this is not a hot code path, I think doing clone on every action would be a reasonable compromise.

Reply:
Your comment above sounds like a nice middle-ground to me. For now, what I have done is to rename `commit_all` to `commit_with` in https://github.com/delta-io/delta-rs/pull/84/commits/e288fc065eb55695cf88676f09f5625390d37667. `commit_all` as a name "sounds" exclusive to me such that callers might expect _all and only actions passed in the call_ to be committed. `commit_with` sounds more inclusive to me as a method that would include actions stored within the tx to be committed along with actions passed to the method. Basically just trying to future-proof the naming a bit.

I agree your suggestion should be pretty straightforward to implement. I would be tempted to implement it immediately, but I feel like it is a semantic I would like to spend more time with outside of laying the foundation.